packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -2927,9 +2930,39 @@ export default (superClass: Class<Parser>): Class<Parser> =>
           this.resetStartLocationFromNode(expr, typeParameters);
         }
         expr.typeParameters = typeParameters;
+
+        // report error if single type parameter used without trailing comma.
+        if (
+          this.hasPlugin("jsx") &&
+          expr.typeParameters.params.length === 1 &&
+          !expr.typeParameters.extra?.trailingComma
+        ) {
+          const parameter = expr.typeParameters.params[0];
+          if (parameter.constraint) {
+            // If parameter has any constraints, it must contain multiple tokens.
+            // <T extends U> is a valid declaration.
+            // <T extends {name: string}> is also a valid declaration.
+          } else {
+            const identifierName = process.env.BABEL_8_BREAKING
+              ? parameter.name.name
+              : parameter.name;
+
+            const start = parameter.start;
+            hasSingleTypeError = { status: true, start, identifierName };
+          }
+        }
+
         return expr;
       }, state);
 
+      if (hasSingleTypeError.status) {
+        this.raise(
+          hasSingleTypeError.start,
```

Comment:
Rather than at `.start`, we could report the error at `parameter.end + 1` which is where we expect the trailing comma.